control-plane/catalog/to-consul/syncer.go

```diff
@@ -173,91 +174,91 @@ func (s *ConsulSyncer) watchReapableServices(ctx context.Context) {
 	// because we have no tracked services in our maps yet.
 	<-s.initialSync
 
+	// Run immediately the first time, then wait for the retry period
+	waitCh := time.After(0)
+	waitBeforeRetry := s.SyncPeriod / 4
+
+	for {
+		select {
+		case <-waitCh:
+			s.deregisterRemovedServices(ctx)
+			waitCh = time.After(waitBeforeRetry)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// deregisterRemovedServices queries the Consul catalog for all services and
+// schedules for deregistration any that no longer have a corresponding k8s
+// service.
+//
+// This function is very similar to [deregisterRemovedService] but handles the case
+// where the ServiceWatcher has been terminated but the service hasn't been deregistered
+// yet.
+func (s *ConsulSyncer) deregisterRemovedServices(ctx context.Context) {
 	opts := &api.QueryOptions{
 		AllowStale: true,
-		WaitIndex:  1,
-		WaitTime:   1 * time.Minute,
 		Filter:     fmt.Sprintf("\"%s\" in Tags", s.ConsulK8STag),
 	}
 
 	if s.EnableNamespaces {
 		opts.Namespace = "*"
 	}
 
-	// minWait is the minimum time to wait between scheduling service deletes.
-	// This prevents a lot of churn in services causing high CPU usage.
-	minWait := s.SyncPeriod / 4
-	minWaitCh := time.After(0)
-	for {
-		// Create a new consul client.
-		consulClient, err := consul.NewClientFromConnMgr(s.ConsulClientConfig, s.ConsulServerConnMgr)
-		if err != nil {
-			s.Log.Error("failed to create Consul API client", "err", err)
-			return
-		}
+	consulClient, err := consul.NewClientFromConnMgr(s.ConsulClientConfig, s.ConsulServerConnMgr)
+	if err != nil {
+		s.Log.Error("failed to create Consul API client", "error", err)
+		return
+	}
 
-		var services *api.CatalogNodeServiceList
-		var meta *api.QueryMeta
-		err = backoff.Retry(func() error {
-			services, meta, err = consulClient.Catalog().NodeServiceList(s.ConsulNodeName, opts)
-			return err
-		}, backoff.WithContext(backoff.NewExponentialBackOff(), ctx))
```

Comment:
This is one of 2 major problems addressed in this PR. The fact that these backoffs had no maximum meant that they would retry forever, until the sync process was terminated.

This combined with using blocking queries which returned errors when nothing happened in the wait period meant that the services were never actually returned in many cases and so expected deregistrations would never happen once a k8s `Service` was deleted.